src/main/java/org/broadinstitute/hellbender/tools/sv/cluster/CanonicalSVLinkage.java

```diff
@@ -0,0 +1,252 @@
+package org.broadinstitute.hellbender.tools.sv.cluster;
+
+import htsjdk.samtools.SAMSequenceDictionary;
+import htsjdk.variant.variantcontext.StructuralVariantType;
+import org.broadinstitute.hellbender.tools.sv.SVCallRecord;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.Utils;
+
+import java.util.Collection;
+import java.util.Set;
+
+/**
+ * <p>Main class for SV clustering. Two items are clustered together if they:</p>
+ * <ul>
+ *   <li>Match event types</li>
+ *   <li>Match strands</li>
+ *   <li>Match contigs</li>
+ *   <li>Meet minimum interval reciprocal overlap (unless inter-chromosomal or a BND).</li>
+ *   <li>Meet minimum sample reciprocal overlap using carrier status. Genotypes (GT fields) are used to determine carrier
+ *   status first. If not found and the event is of type DEL or DUP, then copy number (CN fields) are used instead.
+ *   In the latter case, it is expected that ploidy can be determined from the number of entries in GT.</li>
+ *   <li>Start and end coordinates are within a defined distance.</li>
+ * </ul>
+ *
+ * <p>Interval overlap, sample overlap, and coordinate proximity parameters are defined separately for depth-only/depth-only,
+ * depth-only/PESR, and PESR/PESR item pairs using the {@link ClusteringParameters} class. Note that all criteria must be met for
+ * two candidate items to cluster, unless the comparison is depth-only/depth-only, in which case only one of
+ * interval overlap or break-end proximity must be met. For insertions with undefined length (SVLEN less than 1), interval
+ * overlap is tested assuming the given start position and a length of
+ * {@value CanonicalSVLinkage#INSERTION_ASSUMED_LENGTH_FOR_OVERLAP}.</p>
+ */
+public class CanonicalSVLinkage<T extends SVCallRecord> implements SVClusterLinkage<T> {
+
+    protected final SAMSequenceDictionary dictionary;
+    protected final boolean enableCNV;  // DEL/DUP multi-allelic site clustering
+    protected ClusteringParameters depthOnlyParams;
+    protected ClusteringParameters mixedParams;
+    protected ClusteringParameters evidenceParams;
+
+    public static final int INSERTION_ASSUMED_LENGTH_FOR_OVERLAP = 50;
+
+    public static final double DEFAULT_RECIPROCAL_OVERLAP_DEPTH_ONLY = 0.8;
+    public static final int DEFAULT_WINDOW_DEPTH_ONLY = 0;
+    public static final double DEFAULT_SAMPLE_OVERLAP_DEPTH_ONLY = 0;
+
+    public static final double DEFAULT_RECIPROCAL_OVERLAP_MIXED = 0.8;
+    public static final int DEFAULT_WINDOW_MIXED = 1000;
+    public static final double DEFAULT_SAMPLE_OVERLAP_MIXED = 0;
+
+    public static final double DEFAULT_RECIPROCAL_OVERLAP_PESR = 0.5;
+    public static final int DEFAULT_WINDOW_PESR = 500;
+    public static final double DEFAULT_SAMPLE_OVERLAP_PESR = 0;
+
+    public static final ClusteringParameters DEFAULT_DEPTH_ONLY_PARAMS =
```

Comment:
I decided it was a bit confusing to use the depth-only parameters for defragmentation since reciprocal overlap is not used for the latter, so these parameters in CNVLinkage:
```
    public static final double DEFAULT_SAMPLE_OVERLAP = 0.8;
    public static final double DEFAULT_PADDING_FRACTION = 0.25;
```
are specific to that class.